gflownet/envs/crystals/composition.py

```diff
@@ -258,6 +287,65 @@ def get_element_mask(min_atoms, max_atoms):
+        # the space group and the number of atoms
+        if self.do_spacegroup_check and isinstance(self.space_group, int):
+            space_group = get_space_group(self.space_group)
+            n_atoms = [s for s in state if s > 0]
+
+            # Get the multiplicity of the group's most specific fixed wyckoff position
+            fixed_multiplicity = space_group.get_wyckoff_position(-1).multiplicity
+
+            # Get the multiplicity of the group's most specific wyckoff position with
+            # at least one degree of freedom
+            free_multiplicity = space_group_lowest_free_wp_multiplicity(
+                self.space_group
+            )
+
+            # Go through each action in the masks, validating them
+            # individually
+            for action_idx, nb_atoms_action in enumerate(
+                range(self.min_atom_i, self.max_atom_i + 1)
+            ):
+                if (
+                    not mask_required_element[action_idx]
+                    or not mask_unrequired_element[action_idx]
+                ):
+                    # If the number of atoms added by this action is not a
+                    # multiple of the most specific wyckoff position, mark
+                    # action as invalid
+                    if nb_atoms_action % fixed_multiplicity != 0:
+                        mask_required_element[action_idx] = True
+                        mask_unrequired_element[action_idx] = True
+                        continue
+
+                    # If the number of atoms added by this action is a
+                    # multiple of a non-specific wyckoff position, nothing
+                    # prevents it from being valid
+                    if nb_atoms_action % free_multiplicity == 0:
+                        continue
+
+                    # Checking validity by induction. If a composition is
+                    # valid, adding a number of atoms is equal to the
+                    # multiplicity of a non-specific position, then this
+                    # action must also be valid.
+                    if nb_atoms_action > free_multiplicity and (
+                        not mask_required_element[action_idx - free_multiplicity]
+                        or not mask_unrequired_element[action_idx - free_multiplicity]
+                    ):
+                        continue
+
+                    # If the composition resulting from this action is
+                    # incompatible with the space group, mark action as
+                    # invalid
+                    n_atoms_post_action = n_atoms + [nb_atoms_action]
+                    sg_compatible = space_group_check_compatible(
+                        self.space_group, n_atoms_post_action
+                    )
+                    if not sg_compatible:
+                        mask_required_element[action_idx] = True
+                        mask_unrequired_element[action_idx] = True
```

Comment:
Nice stuff! Suggestion, but as you wish: perhaps pack this into a separate method?

Reply:
Yeah, I agree that this method is starting to be pretty long. I was wondering where I could put the method's boundries so that it is well contained but without having to provide a huge number of arguments. Any suggestion?

Reply:
I have taken a stab at it [here](https://github.com/alexhernandezgarcia/gflownet/commit/48df853d64f8078430df2e05d17955c798772550). What do you think?

Reply:
Seems absolutely fine. I don't know why I thought the result would be messier, this is pretty good.

There are only two things that I'm wondering about : 
- The method since to be both modifying the input masks inplace and returning the updated masks afterwards. Might be simpler to have only one of these mechanism. For speed reason, we might prefer the updating in place. If that's the case, we would need to update the docstring to make it obvious that the change happens inplace.
- In the docstring argument descriptions, it says `Input element-wise mask indicating whether the element is required` for one of the arguments and something similar for the other arguments/return values. That's not exactly right. Each input is an element-wise mask indicating whether each action is invalid for a type of elements. There is one mask for the elements that are required by the environment (they absolutely have to be in the composition) and one mask for the other elements which are not required but may be added if there is room for them.
I'd reword to something like 
```
Input element-wise mask indicating invalid actions for required elements.
This masks indicates whether each individual actions is invalid or not for elements that are required to be in the composition by the end of the trajectory.
```
and
```
Input element-wise mask indicating invalid actions for unrequired elements.
This masks indicates whether each individual actions is invalid or not for elements that are not required to be in the composition by the end of the trajectory.
```

Reply:
Sorry for the errors in the docs and thanks for the suggestions! I have made the changes and opened a [PR](https://github.com/alexhernandezgarcia/gflownet/pull/165).